apps/meteor/server/services/video-conference/service.ts

```diff
@@ -0,0 +1,238 @@
+		this.Messages = new MessagesRaw(db.collection('rocketchat_message'));
+	}
+
+	public async start(caller: IUser['_id'], rid: string): Promise<VideoConferenceInstructions> {
+		const room = await this.Rooms.findOneById<Pick<IRoom, '_id' | 't' | 'uids'>>(rid, { projection: { t: 1, uids: 1 } });
+
+		if (!room) {
+			throw new Error('invalid-room');
+		}
+
+		if (room.t === 'd' && room.uids && room.uids.length <= 2) {
+			return this.startDirect(caller, room);
+		}
+
+		throw new Error('Conferece calls have not been implemented yet.');
+	}
+
+	public async join(uid: IUser['_id'], callId: IVideoConference['_id'], options: VideoConferenceJoinOptions): Promise<string> {
+		const call = await this.VideoConference.findOneById(callId);
+		if (!call) {
+			throw new Error('invalid-call');
+		}
+
+		const user = await this.Users.findOneById<Pick<IUser, '_id' | 'username' | 'name'>>(uid, { projection: { name: 1, username: 1 } });
+		if (!user) {
+			throw new Error('failed-to-load-own-data');
+		}
+
+		switch (call.type) {
+			case 'direct':
+				return this.joinDirect(call, user, options);
+		}
+
+		throw new Error('Conferece calls have not been implemented yet.');
+	}
+
+	public async cancel(uid: IUser['_id'], callId: IVideoConference['_id']): Promise<void> {
+		const call = await this.VideoConference.findOneById(callId);
+		if (!call || !isDirectVideoConference(call)) {
+			throw new Error('invalid-call');
+		}
+
+		if (call.status !== VideoConferenceStatus.CALLING || call.endedBy || call.endedAt) {
+			throw new Error('invalid-call-status');
+		}
+
+		const user = await this.Users.findOneById(uid);
+		if (!user) {
+			throw new Error('failed-to-load-own-data');
+		}
+
+		if (call.messages.started) {
+			await this.changeMessageType(call.messages.started, 'video-direct-missed');
+		}
+
+		this.VideoConference.setEndedById(call._id, { _id: user._id, name: user.name, username: user.username });
+	}
+
+	public async get(callId: IVideoConference['_id']): Promise<IVideoConference | null> {
+		return this.VideoConference.findOneById(callId);
+	}
+
+	private async createMessage(
+		type: MessageTypesValues,
+		rid: IRoom['_id'],
+		user: IUser,
+		extraData: Partial<IMessage> = {},
+	): Promise<IMessage['_id']> {
+		const readReceipt = settings.get('Message_Read_Receipt_Enabled');
+
+		const record = {
+			t: type,
+			rid,
+			ts: new Date(),
+			u: {
+				_id: user._id,
+				username: user.username,
+			},
+			groupable: false,
+			...(readReceipt ? { unread: true } : {}),
+			...extraData,
+		};
+
+		const id = (await this.Messages.insertOne(record)).insertedId;
+		this.Rooms.incMsgCountById(rid, 1);
+		return id;
+	}
+
+	private async createDirectCallMessage(rid: IRoom['_id'], user: IUser): Promise<IMessage['_id']> {
+		return this.createMessage('video-direct-calling', rid, user);
+	}
+
+	private async changeMessageType(messageId: string, newType: MessageTypesValues): Promise<void> {
+		this.Messages.setTypeById(messageId, newType);
```

Comment:
since this method is returning `void`, if we don't `await` here, others awaiting for `changeMessageType` will have no effect.
```suggestion
		await this.Messages.setTypeById(messageId, newType);
```